### PR TITLE
Minor refactoring of `ExchangeReceiver`

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: ref #5900

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by fuzhe1989 — APPROVED** [review]

**Review by guo-shaoge — APPROVED** [review]

File: dbms/src/Flash/Mpp/ExchangeReceiver.cpp [hunk 1]
```diff
@@ -740,56 +725,58 @@ ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::nextResult(std::queue<B
...
+        if (unlikely(!select_resp->ParseFromString(*(recv_msg->resp_ptr))))
+        {
+            return ExchangeReceiverResult::newError(recv_msg->source_index, recv_msg->req_info, "decode error");
+        }
+        else
         {
-            auto select_resp = std::make_shared<tipb::SelectResponse>();
-            if (!select_resp->ParseFromString(*(recv_msg->resp_ptr)))
+            auto result = ExchangeReceiverResult::newOk(select_resp, recv_msg->source_index, recv_msg->req_info);
+            /// If mocking TiFlash as TiDB, here should decode chunks from select_resp.
```
Comment (ywqzzy) [inline dbms/src/Flash/Mpp/ExchangeReceiver.cpp:758]:
```suggestion
            /// If mocking TiFlash as TiDB, we should decode chunks from select_resp.
```

File: dbms/src/Flash/Mpp/ExchangeReceiver.cpp [hunk 2]
```diff
@@ -740,56 +725,58 @@ ExchangeReceiverResult ExchangeReceiverBase<RPCContext>::nextResult(std::queue<B
...
             {
-                result = ExchangeReceiverResult::newError(recv_msg->source_index, recv_msg->req_info, "decode error");
+                assert(recv_msg->chunks.empty());
+                // Fine grained shuffle should only be enabled when sending data to TiFlash node.
+                // So all data should be encoded into MPPDataPacket.chunks.
+                RUNTIME_CHECK_MSG(!enableFineGrainedShuffle(fine_grained_shuffle_stream_count), "Data should not be encoded into tipb::SelectResponse.chunks when fine grained shuffle is enabled");
+                result.decode_detail = CoprocessorReader::decodeChunks(select_resp, block_queue, header, schema);
             }
-            else
+            else if (!recv_msg->chunks.empty())
```
Comment (ywqzzy) [inline dbms/src/Flash/Mpp/ExchangeReceiver.cpp:767]: Is the  check "(!recv_msg->chunks.empty())" necessary?

**Review by ywqzzy — APPROVED** [review]
LGTM

File: dbms/src/Flash/Mpp/ExchangeReceiver.cpp (hunk 2 above)
Comment (SeaRise) [reply dbms/src/Flash/Mpp/ExchangeReceiver.cpp:767]: For the last packet, chunks may be empty.

**Pull request merged**